src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala

```diff
@@ -0,0 +1,109 @@
+import mesosphere.marathon.core.storage.store.{ IdResolver, PersistenceStore }
+import mesosphere.marathon.core.storage.store.impl.zk.{ ZkId, ZkPersistenceStore, ZkSerialized }
+import mesosphere.marathon.storage.migration.MigrationTo146.Environment
+import mesosphere.marathon.storage.repository.InstanceRepository
+import play.api.libs.json.{ JsValue, Json }
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+@SuppressWarnings(Array("ClassNames"))
+class MigrationTo160(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit ctx: ExecutionContext, mat: Materializer) extends StrictLogging {
+
+  def migrate(): Future[Done] = {
+    MigrationTo160.migrateReservations(instanceRepository, persistenceStore)(Environment(sys.env), ctx, mat)
+  }
+}
+
+object MigrationTo160 extends StrictLogging {
+  def migrateReservations(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit env: Environment, ctx: ExecutionContext, mat: Materializer): Future[Done] = {
+
+    logger.info("Starting reservations migration to 1.6.0")
+
+    val maybeStore: Option[ZkPersistenceStore] = {
+
+      def findZkStore(ps: PersistenceStore[_, _, _]): Option[ZkPersistenceStore] = {
+        ps match {
+          case zk: ZkPersistenceStore =>
+            Some(zk)
+          case lcps: LazyCachingPersistenceStore[_, _, _] =>
+            findZkStore(lcps.store)
+          case lvcps: LazyVersionCachingPersistentStore[_, _, _] =>
+            findZkStore(lvcps.store)
+          case ltcps: LoadTimeCachingPersistenceStore[_, _, _] =>
+            findZkStore(ltcps.store)
+          case other =>
+            None
+        }
+      }
+
+      findZkStore(persistenceStore)
+    }
+
+    implicit val instanceResolver: IdResolver[Instance.Id, JsValue, String, ZkId] =
+      new IdResolver[Instance.Id, JsValue, String, ZkId] {
+        override def toStorageId(id: Id, version: Option[OffsetDateTime]): ZkId =
+          ZkId(category, id.idString, version)
+
+        override val category: String = "instance"
+
+        override def fromStorageId(key: ZkId): Id = Instance.Id(key.id)
+
+        override val hasVersions: Boolean = false
+
+        override def version(v: JsValue): OffsetDateTime = OffsetDateTime.MIN
+      }
+
+    implicit val instanceJsonUnmarshaller: Unmarshaller[ZkSerialized, JsValue] =
+      Unmarshaller.strict {
+        case ZkSerialized(byteString) =>
+          Json.parse(byteString.utf8String)
+      }
+
+    import Reservation.reservationFormat
+    import Instance.instanceJsonReads
+
+    maybeStore.map { store =>
+      instanceRepository
+        .ids()
+        .mapAsync(1) { instanceId =>
+          store.get(instanceId)
+        }
+        .mapConcat {
+          case Some(jsValue) =>
+            val instance = jsValue.as[Instance]
+            val maybeReservationJson = (jsValue \ "tasksMap" \\ "reservation").headOption
```

Comment:
I think, it is worth mentioning that prior to 1.6.0, only PVs are supported only with apps, therefore `reservation` can only appear in app instances, and in such of apps, `taskMap` has only one KV pair.

Reply:
Yes, but this code is also handling the case where we don't have any reservation at all, so I'd keep it as is.

Reply:
Yes, I agree. I was not suggestion any change to the code. Instead, I was thinking that it would be a good idea to clarify why it happens the way it happens here in a comment. What do you think?

Reply:
Sounds good!